src/main/scala/strawman/collection/mutable/ListBuffer.scala

```diff
@@ -33,29 +41,163 @@ class ListBuffer[A]
     aliased = false
   }
 
+  private def ensureUnaliased() = if (aliased) copyElems()
+
   /** Convert to list; avoids copying where possible. */
   def toList = {
     aliased = true
     first
   }
 
+  def clear(): Unit = {
+    first = Nil
+  }
+
   def +=(elem: A) = {
-    if (aliased) copyElems()
-    val last1 = elem :: Nil
-    last match {
-      case last: ::[A] => last.next = last1
-      case _ => first = last1
-    }
+    ensureUnaliased()
+    val last1 = (elem :: Nil).asInstanceOf[::[A]]
+    if (len == 0) first = last1 else last.next = last1
     last = last1
     len += 1
     this
   }
 
+  private def locate(i: Int): Predecessor[A] =
+    if (i == 0) null
+    else if (i == len) last
+    else {
+      assert(i > 0 && i < len)
+      var j = i - 1
+      var p = first
+      while (j > 0) {
+        p = p.tail
+        j -= 1
+      }
+      p.asInstanceOf[Predecessor[A]]
+    }
+
+  private def getNext(p: Predecessor[A]): List[A] =
+    if (p == null) first else p.next
+
+  private def setNext(p: Predecessor[A], nx: List[A]): Unit =
+    if (p == null) first = nx else p.next = nx
+
+  def update(idx: Int, elem: A): Unit = {
+    ensureUnaliased()
+    if (idx < 0 || idx >= len) throw new IndexOutOfBoundsException
+    val p = locate(idx)
+    setNext(p, elem :: getNext(p).tail)
+  }
+
+  def insert(idx: Int, elem: A): Unit = {
+    ensureUnaliased()
+    if (idx < 0 || idx > len) throw new IndexOutOfBoundsException
+    if (idx == len) +=(elem)
+    else {
+      val p = locate(idx)
+      setNext(p, elem :: getNext(p))
+      len += 1
+    }
+  }
+
+  private def insertAfter(p: Predecessor[A], it: Iterator[A]) = {
+    var prev = p
+    val follow = getNext(prev)
+    while (it.hasNext) {
+      len += 1
+      val next = (it.next :: follow).asInstanceOf[::[A]]
+      setNext(prev, next)
+      prev = next
+    }
+  }
+
+  def insertAll(idx: Int, elems: IterableOnce[A]): Unit = {
+    ensureUnaliased()
+    val it = elems.iterator()
+    if (it.hasNext) {
+      ensureUnaliased()
+      if (idx < 0 || idx > len) throw new IndexOutOfBoundsException
+      if (idx == len) ++=(elems)
+      else insertAfter(locate(idx), it)
+    }
+  }
+
+  def remove(idx: Int): A = {
+    ensureUnaliased()
+    if (idx < 0 || idx >= len) throw new IndexOutOfBoundsException
+    len -= 1
+    val p = locate(idx)
+    val nx = getNext(p)
+    setNext(p, nx.tail)
+    nx.head
+  }
+
+  def remove(idx: Int, n: Int): Unit =
+    if (n > 0) {
+      ensureUnaliased()
+      if (idx < 0 || idx + n > len) throw new IndexOutOfBoundsException
+      removeAfter(locate(idx), n)
+    }
+
+  private def removeAfter(prev: Predecessor[A], n: Int) = {
+    @tailrec def ahead(p: List[A], n: Int): List[A] =
+      if (n == 0) p else ahead(p.tail, n - 1)
+    setNext(prev, ahead(getNext(prev), n))
+    len -= n
+  }
+
+  def mapInPlace(f: A => A): this.type = {
+    ensureUnaliased()
+    val buf = new ListBuffer[A]
```

Comment:
@Ichoran it is “in place” because at the end the reference returned by the operation is `this` (whereas the usual `map` method returns a new collection).